proposals/extensions/android-parcelable.md

```diff
@@ -0,0 +1,254 @@
+
+```kotlin
+@PaperParcel
+data class User(
+    val id: Long,
+    val firstName: String,
+    val lastName: String
+) : PaperParcelable {
+  companion object {
+    @JvmField val CREATOR = PaperParcelUser.CREATOR
+  }
+}
+```
+
+## Proposal
+
+We can have a compiler extension to generate all the required boilerplate behind the scenes. This is similar to what (some) annotation processors do, but 
+1. Java annotation processors can't alter Kotlin code, 
+2. Normally, annotation processors can't add methods to existing classes (this is possible though a private API only),
+3. A compiler extension can be potentially more flexible in terms of suppressing errors and providing syntactic means to the user.  
+
+### Simple case: completely automatic Parcelable
+
+A compiler extension can generate serialization/deserialization logic for all properties in a primary constructor of a class marked with a special annotation:
+
+```kotlin
+@Parcelize
+class MyParcelable(val data: Int): Parcelable
+```  
+
+Note that the class is not required to be a data class.
+
+The following requirements apply here:
+- The class annotated with `@Parcelize` must implement `Parcelable` (directly or through a chain of supertypes)
+  - Otherwise it's a compiler error
+- All parameters of the primary constructor must be properties, otherwise they can not be (de)serialized
+  - If there is a non-property parameter, it is a compilation error
+- Properties with initializers declared in the class body are also difficult to deserialize correctly: we'd have to generate an alternative constructor that does not execute initializers (including `init` blocks) at all and only uses the serialized data, but this has all the issues that `java.io.Serializable` has wrt "magic" object creation.
+  - Properties in the class body must be marked with a `@Transient` annotation, otherwise it's a compiler warning (or error?) 
+- If some properties are not parcelable (i.e. do not implement `Parcelable`, are not supported by the `Parcel` interface, and are not customized (see below)), it's a compilation error
+- The user is not allowed to manually override methods of `Parcelable` or create the `CREATOR` field
+   - it results in a compilation error  
+   - Question: maybe overriding `describeContents()` is OK?
+
+The annotations for Parcelable classes, transient fields, etc. should sit in a complimentary runtime library that ships together with the compiler extension.  
+
+TODO: Check interactions with inheritance by delegation.
+  
+Some syntactic options:    
+- The names `Parcelize` and `Transient` are just placeholders for now.
+- We could allow the user to omit the supertype `Parcelable`, to be more DRY and have only the annotation to signify that the class is Parcelable, but this would be more challenging wrt the tooling support.
+- We could annotate the supertype itself to make it more local: 
+  - `class MyParcelable(val data: Int): @Auto Parcelable`  
```

Comment:
It's not a matter of line length, it's that the annotation does have an effect on the class, outside of the Parcelable interface. The annotation (not the interface) adds requirements to the class (dealing with property types and constructor parameters) which are not inherited from the interface. Therefore it should be applied to the class since it's modifying that and not the interface, IMHO.